### PR TITLE
Add plugins support

### PR DESCRIPTION
**Pull request by mstemm**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> If contributing rules or changes to rules, please make sure to also uncomment one of the following line:

> /kind rule-update

/kind rule-create

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

/area rules

> /area tests

> /area proposals

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Add support for plugins to extend falco functionality to new event sources/custom fields.
```

### CODE REVIEW

File: userspace/falco/falco.cpp [hunk 1]
```diff
@@ -137,8 +137,8 @@ static void usage()
 	   " -l <rule>                     Show the name and description of the rule with name <rule> and exit.\n"
 	   " --list [<source>]             List all defined fields. If <source> is provided, only list those fields for\n"
 	   "                               the source <source>. Current values for <source> are \"syscall\", \"k8s_audit\"\n"
-	   " --list-plugins                Print info on all loaded plugins and exit.\n"
 #ifndef MINIMAL_BUILD
+	   " --list-plugins                Print info on all loaded plugins and exit.\n"
```
Comment (leogr) [inline userspace/falco/falco.cpp:141]: It seems to me that this commit is not implementing the wanted behavior.

Indeed, `MINIMAL_BUILD` is not the feature gate for the statically linked build. I guess you wanted to use `MUSL_OPTIMIZED_BUILD` instead :point_down: 

```
if(MINIMAL_BUILD)
  set(MINIMAL_BUILD_FLAGS "-DMINIMAL_BUILD")
endif()

if(MUSL_OPTIMIZED_BUILD)
  set(MUSL_FLAGS "-static -Os -fPIE -pie")
endif()
```

right?

**Review by jasondellaluce — COMMENTED** [review]
I left few comments. This looks great!

File: userspace/falco/falco.cpp (hunk 1 above)
Comment (mstemm) [reply userspace/falco/falco.cpp]: Sure, makes sense, I changed it.

File: test/plugins/incompat_api.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+char* plugin_get_required_api_version()
+{
+	return strdup(pl_required_api_version);
+}
+
+extern "C"
+void plugin_free_mem(void *ptr)
+{
+	free(ptr);
+}
```
[the remark this answers is not in the document]
Comment (mstemm) [reply test/plugins/incompat_api.cpp]: Thanks for noticing! Fixed.

File: test/plugins/incompat_api.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+#include <stdlib.h>
+
+// Don't need any function other than plugin_get_required_api_version,
+// plugin load will fail after that.
+static const char *pl_required_api_version = "10000000.0.0";
+
+extern "C"
+char* plugin_get_required_api_version()
+{
+	return strdup(pl_required_api_version);
```
[the remark this answers is not in the document]
Comment (mstemm) [reply test/plugins/incompat_api.cpp]: Indeed! Fixed.

File: test/plugins/test_extract.cpp [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+char* plugin_get_last_error(ss_plugin_t* s)
+{
+	return NULL;
+}
+
+extern "C"
+void plugin_free_mem(void *ptr)
+{
+    free(ptr);
+}
```
[the remark this answers is not in the document]
Comment (mstemm) [reply test/plugins/test_extract.cpp:98]: Yeah I forgot to update this fake plugin after the memory management changes. I fixed this (and the others).

File: test/plugins/test_extract.cpp [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+// This struct represents the state of a plugin. Just has a placeholder string value.
+typedef struct plugin_state
+{
+} plugin_state;
+
+extern "C"
+char* plugin_get_required_api_version()
+{
+	return strdup(pl_required_api_version);
```
[the remark this answers is not in the document]
Comment (mstemm) [reply test/plugins/test_extract.cpp]: done.

File: test/plugins/test_source.cpp [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+} plugin_state;
+
+typedef struct instance_state
+{
+} instance_state;
+
+extern "C"
+char* plugin_get_required_api_version()
+{
+	return strdup(pl_required_api_version);
```
[the remark this answers is not in the document]
Comment (mstemm) [reply test/plugins/test_source.cpp:45]: done.

File: test/plugins/test_source.cpp [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+char* plugin_get_last_error(ss_plugin_t* s)
+{
+	return NULL;
+}
+
+extern "C"
+void plugin_free_mem(void *ptr)
+{
+    free(ptr);
+}
```
[the remark this answers is not in the document]
Comment (mstemm) [reply test/plugins/test_source.cpp:106]: Done.

File: userspace/falco/falco.cpp [hunk 8]
```diff
@@ -916,10 +925,133 @@ int falco_init(int argc, char **argv)
...
+		std::shared_ptr<gen_event_formatter_factory> plugin_formatter_factory(new sinsp_evt_formatter_factory(inspector, plugin_filter_checks));
+
+		std::shared_ptr<sinsp_plugin> input_plugin;
+		std::list<std::shared_ptr<sinsp_plugin>> extractor_plugins;
+		for(auto &p : config.m_plugins)
+		{
+			std::shared_ptr<sinsp_plugin> plugin;
+#ifdef MINIMAL_BUILD
+			falco_logger::log(LOG_INFO, "*Not* loading plugin (" + p.m_name + ") from file " + p.m_library_path + ": Minimal falco does not support plugins.\n");
+			continue;
```
[the remark this answers is not in the document]
Comment (mstemm) [reply userspace/falco/falco.cpp:952]: Good point, I changed it to throw an exception.

File: CMakeLists.txt [hunk 9]
```diff
@@ -220,5 +221,8 @@ add_subdirectory(userspace/engine)
 add_subdirectory(userspace/falco)
 add_subdirectory(tests)
 
+# Reenable once test-infra is publishing plugins artifacts
```
Comment (leogr) [inline CMakeLists.txt:224]: Is the comment out-to-date?

Comment (leogr) [inline CMakeLists.txt:224]: Is the comment out-to-date?

File: test/plugins/test_extract.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (leogr) [reply test/plugins/test_extract.cpp:98]: I still see `plugin_free_mem` here

**Review by leogr — APPROVED** [review]
After much effort, this PR finally looks great!
Thank you very much!

The only thing I don't like is that plugin support doesn't work with the binary package, but I understand the reason. That's something we can see later. I have also left a couple of comments.

Anyway, those are non-blockers for me, so approved! :partying_face: 
/approve

File: CMakeLists.txt (hunk 9 above)
Comment (mstemm) [reply CMakeLists.txt:224]: yes, it's out-of-date, I fixed it.

**Comment by mstemm** [comment]
Ok I also removed the leftover plugin_free_mem reference. Could you take a look again?

File: test/plugins/test_extract.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mstemm) [reply test/plugins/test_extract.cpp]: I removed this one too, thanks for catching it!

**Review by leogr — APPROVED** [review]

**Review by kris-nova — APPROVED** [review]
LGTM

**Pull request merged**